# 01. Importing libraries

In [1]:
# Importlibraries
import pandas as pd
import numpy as np
import os

# 02. Import Data

In [2]:
# import products.csv from Original Data
df_prods = pd.read_csv(r'/Users/bdigregorio/Desktop/CareerFoundry/' \
r'Data Immersion/Data Immersion 4/DI 4.2/Instacart Basket Analysis' \
r'/02 Data/Original Data/products.csv', index_col = False)

In [3]:
path = r'/Users/bdigregorio/Desktop/CareerFoundry/' \
r'Data Immersion/Data Immersion 4/DI 4.2/Instacart Basket Analysis'

In [4]:
path

'/Users/bdigregorio/Desktop/CareerFoundry/Data Immersion/Data Immersion 4/DI 4.2/Instacart Basket Analysis'

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False) 

# 03. Consistency Check

In [7]:
# Create a dataframe

df_test = pd.DataFrame()

In [8]:
# Create a mixed type column

df_test['mix']= ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
df_test['mix'] = df_test['mix'].astype('str')

In [12]:
# check for mixed types again

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [13]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# addressing NaNs

# FINDING product_name NaNs with groupby

In [16]:
df_prods.groupby('product_id')['product_name'].unique()

product_id
1                             [Chocolate Sandwich Cookies]
2                                       [All-Seasons Salt]
3                   [Robust Golden Unsweetened Oolong Tea]
4        [Smart Ones Classic Favorites Mini Rigatoni Wi...
5                              [Green Chile Anytime Sauce]
                               ...                        
49684          [Vodka, Triple Distilled, Twist of Vanilla]
49685                 [En Croute Roast Hazelnut Cranberry]
49686                                   [Artisan Baguette]
49687         [Smartblend Healthy Metabolism Dry Cat Food]
49688                             [Fresh Foaming Cleanser]
Name: product_name, Length: 49686, dtype: object

In [17]:
df_prod_id_name = df_prods.groupby('product_id')['product_name'].unique()

In [18]:
df_prod_id_name.head(35)

product_id
1                          [Chocolate Sandwich Cookies]
2                                    [All-Seasons Salt]
3                [Robust Golden Unsweetened Oolong Tea]
4     [Smart Ones Classic Favorites Mini Rigatoni Wi...
5                           [Green Chile Anytime Sauce]
6                                        [Dry Nose Oil]
7                      [Pure Coconut Water With Orange]
8                   [Cut Russet Potatoes Steam N' Mash]
9                   [Light Strawberry Blueberry Yogurt]
10     [Sparkling Orange Juice & Prickly Pear Beverage]
11                                  [Peach Mango Juice]
12                         [Chocolate Fudge Layer Cake]
13                                  [Saline Nasal Mist]
14                     [Fresh Scent Dishwasher Cleaner]
15                           [Overnight Diapers Size 6]
16                      [Mint Chocolate Flavored Syrup]
17                                  [Rendered Duck Fat]
18                [Pizza for One Supr

In [19]:
print(df_prod_id_name.loc[[34,69,116,262,525,1511,1780,2240,2586,3159,3230,3736,4283,4790,38183,40440]])

product_id
34       [nan]
69       [nan]
116      [nan]
262      [nan]
525      [nan]
1511     [nan]
1780     [nan]
2240     [nan]
2586     [nan]
3159     [nan]
3230     [nan]
3736     [nan]
4283     [nan]
4790     [nan]
38183    [nan]
40440    [nan]
Name: product_name, dtype: object


#  NaN exploration summary
After exploring several NaN and found these prouduct_id's with NaNa for product names are
not showing up in other departments or aisles. I believe it is justifiable to remove these 16 rows from the 49693 in the data set.


In [20]:
# contimuing with 4.5 reading exercise
df_prods.shape

(49693, 5)

In [21]:
# removing the NaNs
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [22]:
df_prods_clean.shape

(49677, 5)

In [23]:
# finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [24]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [25]:
df_prods_clean.shape

(49677, 5)

In [26]:
# removing duplicates 
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [27]:
df_prods_clean_no_dups.shape

(49672, 5)

# 04. Exporting df_prods_clean_no_dups

In [28]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 05. 4.5 TASK

In [29]:
# Task pre-steps
# Importing orders_wrangled data

In [30]:
df_orders_wrangled = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [31]:
# Task pre-step 
# Importing products_checked data

In [32]:
df_products_checked = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

In [33]:
# task step 1: above from exercise reading

In [34]:
# task step 2: Run the df.describe() function on your df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# step 2 output
there seems to be something strange about Unnamed: 0. I can't tell what this means. Everything else seems to be fine - except the long exponential numbers (e+)

In [35]:
#step 3 Check for mixed-type data in your df_ords dataframe.

for col in df_ords.columns.tolist(): 
    wierd = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_ords[wierd]) > 0: 
        print (col)

# step 3 output
running the check for mixed data in columns got no result, meaning there is NO column with mixed data in df_ords.

In [36]:
# no need for step 4: If you find mixed-type data, fix it

In [37]:
#step 5 Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [38]:
# step 6 Address the missing values in days_since_prior_order

In [39]:
# reviewing df_ords
df_ords.head(10)

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           0   2539329        1    prior             1          2   
1           1   2398795        1    prior             2          3   
2           2    473747        1    prior             3          3   
3           3   2254736        1    prior             4          4   
4           4    431534        1    prior             5          4   
5           5   3367565        1    prior             6          2   
6           6    550135        1    prior             7          1   
7           7   3108588        1    prior             8          1   
8           8   2295261        1    prior             9          1   
9           9   2550362        1    prior            10          4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0  
5                  7                    19.0  
6                  9                    20.0  
7                 14                    14.0  
8                 16                     0.0  
9                  8                    30.0

In [40]:
# removing unnecessary columns 'Unnamed' and eval_set
df_ords.drop(columns = ['eval_set', 'Unnamed: 0'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [41]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [42]:
df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
11               11   2168274        2    prior             1          2   
26               26   1374495        3    prior             1          1   
39               39   3343014        4    prior             1          6   
45               45   2717275        5    prior             1          3   
...             ...       ...      ...      ...           ...        ...   
3420930     3420930    969311   206205    prior             1          4   
3420934     3420934   3189322   206206    prior             1          3   
3421002     3421002   2166133   206207    prior             1          6   
3421019     3421019   2227043   206208    prior             1          1   
3421069     3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 8 columns]

# step 6 output
I did research on a few user_id's and noticed a pattern. The first order for each customer is NaN: meaning this was their first order so there could not have been a "prior" order. The count on NaNs in this column is 206209 which tells us Instacart has attracted many customers.

In [43]:
# step 7 Run a check for duplicate values in your df_ords data
df_dups = df_ords[df_ords.duplicated()]

In [44]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

#  step 7 output
I used the function to find duplicates and none were found

In [45]:
# step 8 not needed

In [46]:
df_ords.shape

(3421083, 8)

# 06. EXPORTING DATA

In [47]:
# exporting df_ords 
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))

In [48]:
# exporting df_prods 
df_products_checked.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'prods_cleaned.csv'))